metaserver/src/main/scala/org/langmeta/lsp/Endpoints.scala

```diff
@@ -0,0 +1,83 @@
+    notification[DidCloseTextDocumentParams]("textDocument/didClose")
+  val didSave: Endpoint[DidSaveTextDocumentParams, Unit] =
+    notification[DidSaveTextDocumentParams]("textDocument/didSave")
+  val didOpen: Endpoint[DidOpenTextDocumentParams, Unit] =
+    notification[DidOpenTextDocumentParams]("textDocument/didOpen")
+  val didChange: Endpoint[DidChangeTextDocumentParams, Unit] =
+    notification[DidChangeTextDocumentParams]("textDocument/didChange")
+  val completion: Endpoint[TextDocumentPositionParams, CompletionList] =
+    request[TextDocumentPositionParams, CompletionList]("textDocument/completion")
+  val definition: Endpoint[TextDocumentPositionParams, List[Location]] =
+    request[TextDocumentPositionParams, List[Location]]("textDocument/definition")
+  val references: Endpoint[ReferenceParams, List[Location]] =
+    request[ReferenceParams, List[Location]]("textDocument/references")
+  val hover: Endpoint[TextDocumentPositionParams, Hover] =
+    request[TextDocumentPositionParams, Hover]("textDocument/hover")
+  val codeAction: Endpoint[CodeActionParams, List[Command]] =
+    request[CodeActionParams, List[Command]]("textDocument/codeAction")
+  val documentHighlight: Endpoint[TextDocumentPositionParams, List[DocumentHighlight]] =
+    request[TextDocumentPositionParams, List[DocumentHighlight]]("textDocument/documentHighlight")
+  val documentSymbol: Endpoint[DocumentSymbolParams, List[SymbolInformation]] =
+    request[DocumentSymbolParams, List[SymbolInformation]]("textDocument/documentSymbol")
+  val formatting: Endpoint[DocumentFormattingParams, List[TextEdit]] =
+    request[DocumentFormattingParams, List[TextEdit]]("textDocument/formatting")
+  val rename: Endpoint[RenameParams, WorkspaceEdit] =
+    request[RenameParams, WorkspaceEdit]("textDocument/rename")
+  val signatureHelp: Endpoint[TextDocumentPositionParams, SignatureHelp] =
+    request[TextDocumentPositionParams, SignatureHelp]("textDocument/signatureHelp")
+}
+
+object Workspace extends Workspace
+trait Workspace {
+  val didChangeConfiguration: Endpoint[DidChangeConfigurationParams, Unit] =
+    notification[DidChangeConfigurationParams]("workspace/didChangeConfiguration")
+  val didChangeWatchedFiles: Endpoint[DidChangeWatchedFilesParams, Unit] =
+    notification[DidChangeWatchedFilesParams]("workspace/didChangeWatchedFiles")
+  val executeCommand: Endpoint[ExecuteCommandParams, Json] =
+    request[ExecuteCommandParams, Json]("workspace/executeCommand")
+  val symbol: Endpoint[WorkspaceSymbolParams, List[SymbolInformation]] =
+    request[WorkspaceSymbolParams, List[SymbolInformation]]("workspace/symbol")
+  val applyEdit: Endpoint[ApplyWorkspaceEditParams, ApplyWorkspaceEditResponse] =
+    request[ApplyWorkspaceEditParams, ApplyWorkspaceEditResponse]("workspace/applyEdit")
+}
+
+object Window extends Window
+trait Window {
+  object showMessage extends Endpoint[ShowMessageParams, Unit]("window/showMessage") {
```

Comment:
making endpoints objects seems to work quite nicely, I'm tempted to change all the vals to objects for the same benefit. Any thoughts?

Reply:
Saves keystrokes and prevents eyeball strain. 👍